lib/internal/dns/promises.js

```diff
@@ -0,0 +1,249 @@
+'use strict';
+const {
+  bindDefaultResolver,
+  Resolver: CallbackResolver,
+  validateHints
+} = require('internal/dns/utils');
+const { codes, dnsException } = require('internal/errors');
+const { isIP, isIPv4, isLegalPort } = require('internal/net');
+const {
+  getaddrinfo,
+  getnameinfo,
+  ChannelWrap,
+  GetAddrInfoReqWrap,
+  GetNameInfoReqWrap,
+  QueryReqWrap
+} = process.binding('cares_wrap');
+const {
+  ERR_INVALID_ARG_TYPE,
+  ERR_INVALID_OPT_VALUE,
+  ERR_MISSING_ARGS,
+  ERR_SOCKET_BAD_PORT
+} = codes;
+
+
+function onlookup(err, addresses) {
+  if (err) {
+    this.reject(dnsException(err, 'getaddrinfo', this.hostname));
+    return;
+  }
+
+  const family = this.family ? this.family : isIPv4(addresses[0]) ? 4 : 6;
+  this.resolve({ address: addresses[0], family });
+}
+
+function onlookupall(err, addresses) {
+  if (err) {
+    this.reject(dnsException(err, 'getaddrinfo', this.hostname));
+    return;
+  }
+
+  const family = this.family;
+
+  for (var i = 0; i < addresses.length; i++) {
+    const address = addresses[i];
+
+    addresses[i] = {
+      address,
+      family: family ? family : isIPv4(addresses[i]) ? 4 : 6
+    };
+  }
+
+  this.resolve(addresses);
+}
+
+function createLookupPromise(family, hostname, all, hints, verbatim) {
```

Comment:
I would really prefer it if this method dind't do so much explicit construction:

```js
async function createLookupPromise(family, hostname, all, hints, verbatim) {
  if (!hostname) {
    if (all) return [];
   else 
     return { address: null, family: family === 6 ? 6 : 4 };
  }
  const matchedFamily = isIP(hostname);
  if (matchedFamily !== 0) {
    const result = { address: hostname, family: matchedFamily }
    if(all) return [result];
    return result;
  }
  return await new Promise((resolve, reject) => {
    const req = new GetAddrInfoReqWrap();
    // ...
    req.resolve = resolve;
  }
}
```